### PR TITLE
SSZ and Merkleization wiki pages

### PR DESCRIPTION
**Pull request by thogiti**
## Wiki PR Checklist

- [x] SSZ wiki page
- [x] Merkleization wiki page

### CODE REVIEW

**Comment by raxhvl** [comment]
I'll review this after our call today.

File: docs/wiki/CL/merkleization.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
+# Merkleization and Hash Tree Roots 
+
+In Ethereum consensus mechanism, it's critical for all participating nodes to agree on the state of the system consistently and efficiently. The [Simple Serialize (SSZ)](/docs/wiki/CL/SSZ.md) framework facilitates this through Merkleization, a process that transforms serialized data into a Merkle tree structure. This wiki page discusses the intricacies of Merkleization and its importance in ensuring a shared state across nodes in a scalable and secure manner.
```
Comment (taxmeifyoucan) [inline docs/wiki/CL/merkleization.md:3]: I would add that the Merkleization scheme allows access to beacon state proofs for constrained environments like light clients

File: docs/wiki/CL/merkleization.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+
+- **Chunking:** Divide the serialized data into 32-byte chunks.
+- **Tree Construction:** Pair up the chunks and hash each pair to form the next level of the tree. Repeat this step until only one hash remains: the Merkle root.
+- **Padding:** If the number of chunks isn't a power of two, additional zero-value chunks are added to round out the tree, ensuring that the tree is balanced.
+
+## Benefits of Merkleization
+
+- **Performance Efficiency:** While the tree requires hashing approximately twice the original data amount, caching mechanisms can store the roots of subtrees that don't change often. This significantly reduces the computational overhead as only altered parts of the data need re-hashing.
+- **Light Client Support:** The Merkle tree structure supports the creation of Merkle proofs—small pieces of data that prove the inclusion and integrity of specific parts of the state without needing the entire dataset. This feature is crucial for light clients, which operate with limited resources and rely on these proofs to interact with  Ethereum securely.
```
Comment (taxmeifyoucan) [inline docs/wiki/CL/merkleization.md:25]:
```suggestion
- **Light Client Support:** The Merkle tree structure supports the creation of Merkle proofs—small pieces of data that prove the inclusion and integrity of specific parts of the state without needing the entire dataset. This feature is crucial for light clients, which operate with limited resources and rely on these proofs to interact with Ethereum securely.
```

File: docs/wiki/CL/merkleization.md [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+
+## Merkle Tree Structure and Hashing
```
Comment (taxmeifyoucan) [inline docs/wiki/CL/merkleization.md:27]: I am not sure if we need to cover merkle tree basics here. It's also mentioned on EL/data-structures and maybe you can just reference an external resource to keep the article shorter. 
We wrote a merkle tree primer for Summer of protocol, I will link it when it's published

File: docs/wiki/CL/merkleization.md [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+
+There are helper functions in the consensus specs to calculate the multiproofs and generalized indices. You can find [here.](https://github.com/ethereum/consensus-specs/blob/dev/ssz/merkle-proofs.md#merkle-multiproofs)
+
+## Calculating Hash Tree Roots
+
+The hash tree root of an SSZ object is computed recursively. For basic types and collections of basic types, the data is packed into chunks and directly Merkleized. For composite types like containers, the process involves hashing the tree roots of each component. In the below sections we will see the working examples to understand the process.
+
+### Packing and Chunking
+
+Packing and chunking are crucial concepts in the context of Merkleization, especially when dealing with the SSZ used in the Beacon chain. Here's how packing and chunking work:
```
Comment (taxmeifyoucan) [inline docs/wiki/CL/merkleization.md:135]:
```suggestion
Packing and chunking enable the Merkleization with SSZ by formatting the serialized data and dividing them into pieces which are then hashed to a Merkle tree. Here's how the process works:
```

**Review by taxmeifyoucan — APPROVED** [review]

**Comment by thogiti** [comment]
@taxmeifyoucan This is ready to be merged. Thx.

**Comment by taxmeifyoucan** [comment]
Thanks @thogiti! I lgtm, I was just waiting whether @raxhvl or @angaz add a review. Let's give them a bit more time, otherwise I'll merge

File: wordlist.txt [hunk 5]
```diff
@@ -636,3 +636,38 @@ zkEVMs
 ZKSNARK
 ZKSNARKs
 Zksync
+CLRS
```
Comment (angaz) [inline wordlist.txt:639]: Please keep the list sorted.

File: docs/wiki/CL/SSZ.md [hunk 6]
```diff
@@ -0,0 +1,823 @@
...
+## SSZ VS RLP Serialization
+
+| CRITERIA   | COMPACT | EXPRESSIVENESS | HASHING  | INDEXING |
+|------------|---------|----------------|----------|----------|
+| RLP        | Yes     | Flexible       | Possible | No       |
+| SSZ        | No      | Yes            | Yes      | Poor     |
+
+_Table: SSZ VS RLP Comparison by [Piper Merriam](https://twitter.com/pipermerriam)._
+
+1. **Expressiveness**:
```
Comment (angaz) [inline docs/wiki/CL/SSZ.md:29]: Normally we use `1.` for all the ordered lists items, and the markdown renderer will set the correct index. Makes it easier if you need to reorder something or add something in the middle, not needing to renumber everything.

**Review by angaz — COMMENTED** [review]
This is awesome! So much information. So much detail. It's next level.

I just have a few nitpicks.

I'm not sure if we have a policy on line length. I generally go with 80-120. A formatter can probably fix that if it's something we would like to have. Not the end of the world. Just nice for someone who doesn't have wrapping on their editor.

It looks like the colour scheme for the charts don't really work well with dark mode. I'm not sure what to do about that. Maybe just use outlines? Or specify the text colour as well so it isn't converted into a lighter colour in dark mode.
image

File: wordlist.txt (hunk 5 above)
Comment (thogiti) [reply wordlist.txt]: Thank you @angaz for the review and feedback. 

I fixed for these suggestions.

**Comment by raxhvl** [comment]
@thogiti I submited the review after the merge. Small nits. It's a great article overall. Thank you.

**Comment by taxmeifyoucan** [comment]
Sorry for merging too soon! Please push more updates to `cl-architecture` branch and I will merge it

File: wordlist.txt
Comment (angaz) [inline wordlist.txt:1]: I see there's some duplicates in this file. Probably would be nice if we ran it through `uniq`.

File: wordlist.txt [hunk 7]
```diff
@@ -249,6 +266,8 @@ getPayloadV
 getters
 ghost
 Gilles
+Goeril
```
Comment (angaz) [inline wordlist.txt:289]: Kinda sus :)

File: wordlist.txt [hunk 8]
```diff
@@ -419,6 +445,8 @@ PeerDAS
 Peeters
 pepc
 pepc's
+performace
```
Comment (angaz) [inline wordlist.txt:480]: More sus

File: wordlist.txt
Comment (thogiti) [reply wordlist.txt:1]: I think the file from main got merged into the branch file and but didn't correctly dedupe it. 

I will update the fix and remove duplicates in my new PR.

**Comment by thogiti** [comment]
> @thogiti I submited the review after the merge. Small nits. It's a great article overall. Thank you.

Love it. Our wiki will be better with more reviews, feedback and comments. Thank you @raxhvl, @angaz, and @taxmeifyoucan for the feedback.

I will apply the changes/updates in a new PR and request you to merge it.

File: wordlist.txt (hunk 7 above)
Comment (thogiti) [reply wordlist.txt:289]: I couldn't locate any docs that has these misspelled words. Perhaps, someone put it there to get around the spellcheck but then later corrected the grammar in the page.

**Pull request merged**